### PR TITLE
Add RestfulApiWebService solution to tools folder

### PR DESCRIPTION
**Pull request by hechaochao**

### CODE REVIEW

File: tools/RestfulApiWebService/DB/Models/data.Context.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+// </auto-generated>
+//------------------------------------------------------------------------------
+
+namespace DB.Models
+{
+    using System;
+    using System.Data.Entity;
+    using System.Data.Entity.Infrastructure;
+    
+    public partial class dataEntities : DbContext
```
Comment (vwxyzh) [inline tools/RestfulApiWebService/DB/Models/data.Context.cs]: naming

File: tools/RestfulApiWebService/OperationOnDB/Global.asax.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+using System.Web.SessionState;
+using System.Web.Http;
+
+namespace OperationOnDB
+{
+    public class Global : HttpApplication
+    {
+        void Application_Start(object sender, EventArgs e)
+        {
+            // 在应用程序启动时运行的代码
```
Comment (vwxyzh) [inline tools/RestfulApiWebService/OperationOnDB/Global.asax.cs]: ...

**Comment by vwxyzh** [comment]
too many garbage files.

File: tools/RestfulApiWebService/DB/Models/uidt.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+//     Manual changes to this file will be overwritten if the code is regenerated.
+// </auto-generated>
+//------------------------------------------------------------------------------
+
+namespace DB.Models
+{
+    using System;
+    using System.Collections.Generic;
+    
+    public partial class uidt
```
Comment (vwxyzh) [inline tools/RestfulApiWebService/DB/Models/uidt.cs]: >uidt [](start = 25, length = 4)

naming

Comment (vicancy) [inline tools/RestfulApiWebService/DB/Models/uidt.cs]: Please go through https://github.com/dotnet/docfx/blob/dev/Documentation/guideline/csharp_coding_standards.md

File: tools/RestfulApiWebService/DB/createDB.sql [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+GO
+
+/****** Object:  Table [dbo].[uidt]    Script Date: 7/21/2017 9:13:33 PM ******/
+SET ANSI_NULLS ON
+GO
+
+SET QUOTED_IDENTIFIER ON
+GO
+
+CREATE TABLE [dbo].[uidt](
```
Comment (vwxyzh) [inline tools/RestfulApiWebService/DB/createDB.sql]: >uidt [](start = 20, length = 4)

naming

File: tools/RestfulApiWebService/DB/createDB.sql [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+	[uid] [varchar](800) NOT NULL,
+	[objectStr] [varchar](max) NOT NULL,
```
Comment (vwxyzh) [inline tools/RestfulApiWebService/DB/createDB.sql]: >varchar [](start = 14, length = 7)

nvarchar?

File: tools/RestfulApiWebService/DB/createDB.sql [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+/****** Object:  Table [dbo].[uidt]    Script Date: 7/21/2017 9:13:33 PM ******/
+SET ANSI_NULLS ON
+GO
+
+SET QUOTED_IDENTIFIER ON
+GO
+
+CREATE TABLE [dbo].[uidt](
+	[uid] [varchar](800) NOT NULL,
```
Comment (vwxyzh) [inline tools/RestfulApiWebService/DB/createDB.sql]: >varchar [](start = 8, length = 7)

nvarchar?

File: tools/RestfulApiWebService/DB/Dao/XrefSpecDBContext.cs [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+namespace DB.Dao
+{
+    public class XrefSpecDBContext: DbContext
+    {
+        public XrefSpecDBContext(): base("name=dataEntities")
+        {
+
+        }
+
+        public DbSet<XRefSpecObject> uidTable { get; set; }
```
Comment (vwxyzh) [inline tools/RestfulApiWebService/DB/Dao/XrefSpecDBContext.cs]: >uidTable [](start = 37, length = 8)

naming?

File: tools/RestfulApiWebService/DB/Models/XRefSpecObject.cs [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace DB.Models
+{   
+    [Table("uidt",Schema ="dbo")]
+    public class XRefSpecObject
+    {   
+        [Column(TypeName ="varchar")]
+        [MaxLength(900)]
+        [Key]
+        public string uid { get; set; }
```
Comment (vwxyzh) [inline tools/RestfulApiWebService/DB/Models/XRefSpecObject.cs]: >uid [](start = 22, length = 3)

naming?

File: tools/RestfulApiWebService/DB/Models/XRefSpecObject.cs [hunk 9]
```diff
@@ -0,0 +1,17 @@
+﻿using System.ComponentModel.DataAnnotations;
+using System.ComponentModel.DataAnnotations.Schema;
+
+namespace DB.Models
+{   
+    [Table("uidt",Schema ="dbo")]
+    public class XRefSpecObject
+    {   
+        [Column(TypeName ="varchar")]
+        [MaxLength(900)]
```
Comment (vwxyzh) [inline tools/RestfulApiWebService/DB/Models/XRefSpecObject.cs]: >900 [](start = 19, length = 3)

800?

**Comment by vwxyzh** [comment]
OperationOnDB?

File: tools/RestfulApiWebService/OperationOnDB/Filters/ApiBasicAuthenticationAttribute.cs [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+            this.Username = username;
+            this.Password = password;
+        }
+
+        public override void OnActionExecuting(HttpActionContext actionContext)
+        {
+          
+                
+            
+            if(actionContext.Request.Headers.Authorization != null)
```
Comment (vwxyzh) [inline tools/RestfulApiWebService/OperationOnDB/Filters/ApiBasicAuthenticationAttribute.cs]: remove blank lines.

File: tools/RestfulApiWebService/OperationOnDB/Controllers/ApiFileController.cs [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+        {
+
+            XRefMap xref = Microsoft.DocAsCode.Common.YamlUtility.Deserialize<XRefMap>(path);
+            foreach (var spec in xref.References)
+            {
+                XRefSpecObject t = new XRefSpecObject();
+                t.uid = spec["uid"];
+                t.objectStr = Newtonsoft.Json.JsonConvert.SerializeObject(spec);
+                db.uidTable.Add(t);
+                db.SaveChanges();
```
Comment (vicancy) [inline tools/RestfulApiWebService/OperationOnDB/Controllers/ApiFileController.cs]: move outside the loop to batch?

File: tools/RestfulApiWebService/OperationOnDB/Filters/ApiBasicAuthenticationAttribute.cs [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+            {
+                string authToken = actionContext.Request.Headers.Authorization.Parameter;
+                string decodedToken = System.Text.Encoding.UTF8.GetString(Convert.FromBase64String(authToken));
```
Comment (vwxyzh) [inline tools/RestfulApiWebService/OperationOnDB/Filters/ApiBasicAuthenticationAttribute.cs]: >System.Text.Encoding. [](start = 38, length = 21)

using System.Text.Encoding;

File: tools/RestfulApiWebService/OperationOnDB/Filters/ApiBasicAuthenticationAttribute.cs [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+                string username = decodedToken.Substring(0, decodedToken.IndexOf(":"));
+                string password = decodedToken.Substring(decodedToken.IndexOf(":") + 1);
+                if (username == Username && password == Password) return;
```
Comment (vwxyzh) [inline tools/RestfulApiWebService/OperationOnDB/Filters/ApiBasicAuthenticationAttribute.cs]: add {}

File: tools/RestfulApiWebService/RestfulApiWebService/Controllers/QueryController.cs [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+
+        [HttpGet]
+        [Route("{uid}")]
+        public async Task<IHttpActionResult> GetByUid(string uid)
+        {
+            var ut = await db.uidTable.Where(b => b.uid == uid)
+                                .Select(c => c.objectStr)
+                                .ToListAsync();
+           
+            List<XRefSpec> xf = Newtonsoft.Json.JsonConvert.DeserializeObject<List<XRefSpec>>("[" + string.Join(",", ut) + "]");
```
Comment (vwxyzh) [inline tools/RestfulApiWebService/RestfulApiWebService/Controllers/QueryController.cs]: [https://stackoverflow.com/questions/9541351/returning-binary-file-from-controller-in-asp-net-web-api](https://stackoverflow.com/questions/9541351/returning-binary-file-from-controller-in-asp-net-web-api)

File: tools/RestfulApiWebService/RestfulApiWebService/Controllers/QueryController.cs [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+                getUidTasks.Add(tempDB.uidTable.Where(b => b.uid == uid)
+                               .Select(c => c)
+                               .FirstOrDefaultAsync());
+            }
+
+            var uidList = await Task.WhenAll(getUidTasks);
+            foreach(var temp in uidList)
+            {
+                XRefSpec xf = null;
+                if (temp != null) xf = Newtonsoft.Json.JsonConvert.DeserializeObject<XRefSpec>(temp.objectStr);
```
Comment (vwxyzh) [inline tools/RestfulApiWebService/RestfulApiWebService/Controllers/QueryController.cs]: add {}

File: tools/RestfulApiWebService/RestfulApiWebService/Controllers/QueryController.cs [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+           
+            List<XRefSpec> xf = Newtonsoft.Json.JsonConvert.DeserializeObject<List<XRefSpec>>("[" + string.Join(",", ut) + "]");
+            return Ok(xf);
+        }
+           
+        [HttpPost]
+        [Route("")]
+        public async Task<IHttpActionResult> PostByUids([FromBody]string[] uids)
+        {
+            //Stopwatch stopWatch = new Stopwatch();
```
Comment (vwxyzh) [inline tools/RestfulApiWebService/RestfulApiWebService/Controllers/QueryController.cs]: remove.

File: tools/RestfulApiWebService/OperationOnDB/Global.asax.cs (hunk 2 above)
Comment (hechaochao) [inline tools/RestfulApiWebService/OperationOnDB/Global.asax.cs]: remove this comment? which is produced automatically

---
In reply to: [129222616](https://github.com/dotnet/docfx/pull/1906#discussion_r129222616) [](ancestors = 129222616)

File: tools/RestfulApiWebService/DB/Utility/MD5Encryption.cs [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace DB.Utility
+{
+    public class MD5Encryption
+    {
+        public static string CalculateMD5Hash(string input)
+        {
+            // step 1, calculate MD5 hash from input
+            MD5 md5 = MD5.Create();
+            byte[] inputBytes = Encoding.ASCII.GetBytes(input);
```
Comment (vwxyzh) [inline tools/RestfulApiWebService/DB/Utility/MD5Encryption.cs]: >ASCII [](start = 41, length = 5)

utf8/unicode?

File: tools/RestfulApiWebService/DB/createDB.sql (hunk 4 above)
Comment (vwxyzh) [inline tools/RestfulApiWebService/DB/createDB.sql]: out of date

**Comment by superyyrrzz** [comment]
> 

can merge with .gitignore under root?

---
Refers to: tools/RestfulApiWebService/.gitignore:3 in dbe62d2. [](commit_id = dbe62d24491eca7c403a09cd4a5db26a532e09b8, deletion_comment = False)

File: tools/RestfulApiWebService/DB/DB.csproj [hunk 18]
```diff
@@ -0,0 +1,138 @@
+﻿<Project ToolsVersion="12.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
Comment (vicancy) [inline tools/RestfulApiWebService/DB/DB.csproj]: use asp.net core project

File: tools/RestfulApiWebService/DB/Dao/XrefSpecDBContext.cs [hunk 19]
```diff
@@ -0,0 +1,15 @@
+﻿using DB.Models;
+using System.Data.Entity;
+
+namespace DB.Dao
```
Comment (vicancy) [inline tools/RestfulApiWebService/DB/Dao/XrefSpecDBContext.cs]: naming: what does `Dao` mean?

File: tools/RestfulApiWebService/DB/Properties/AssemblyInfo.cs [hunk 20]
```diff
@@ -0,0 +1,35 @@
+﻿using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
```
Comment (vicancy) [inline tools/RestfulApiWebService/DB/Properties/AssemblyInfo.cs]: reuse `src\Shared\base.props` as other projects do and remove this file

File: tools/RestfulApiWebService/DB/Utility/MD5Encryption.cs [hunk 21]
```diff
@@ -0,0 +1,28 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Cryptography;
+using System.Text;
+using System.Web;
+
+namespace DB.Utility
```
Comment (vicancy) [inline tools/RestfulApiWebService/DB/Utility/MD5Encryption.cs]: naming: `DB` to `Database`, no need to abbreviate

File: tools/RestfulApiWebService/DB/Web.Debug.config [hunk 22]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="utf-8"?>
+
```
Comment (vicancy) [inline tools/RestfulApiWebService/DB/Web.Debug.config]: Remove this file

File: tools/RestfulApiWebService/DB/Web.Release.config [hunk 23]
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (vicancy) [inline tools/RestfulApiWebService/DB/Web.Release.config]: Remove this file

File: tools/RestfulApiWebService/DB/createDB.sql [hunk 24]
```diff
@@ -0,0 +1,19 @@
+USE [test]
```
Comment (vicancy) [inline tools/RestfulApiWebService/DB/createDB.sql]: use a subfolder `scripts` for `.sql` files

File: tools/RestfulApiWebService/OperationOnDB/Controllers/ApiFileController.cs [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+            }
+            else
+            {
+                return "Uploaded Failed";
+            }
+        }
+
+        private void SaveData(string path)
+        {
+
```
Comment (vicancy) [inline tools/RestfulApiWebService/OperationOnDB/Controllers/ApiFileController.cs]: remove extra empty lines

File: tools/RestfulApiWebService/OperationOnDB/Controllers/ApiFileController.cs [hunk 26]
```diff
@@ -0,0 +1,75 @@
...
+using Microsoft.DocAsCode.Build.Engine;
+using Microsoft.DocAsCode.Common;
+using Newtonsoft.Json;
+using OperationOnDB.Filters;
+using System.IO;
+using System.Web;
+using System.Web.Hosting;
+using System.Web.Http;
+
+namespace OperationOnDB.Controllers
```
Comment (vicancy) [inline tools/RestfulApiWebService/OperationOnDB/Controllers/ApiFileController.cs]: naming: `OperationOnDB` to `XRefService`?

File: tools/RestfulApiWebService/OperationOnDB/Controllers/ApiFileController.cs [hunk 27]
```diff
@@ -0,0 +1,75 @@
...
+{
+    [ApiBasicAuthenticationAttribute("user", "user",
+    BasicRealm = "localhost")]
+    [RoutePrefix("uids")]
+    public class ApiFileController : ApiController
+    {
+        private XrefSpecDBContext _db = new XrefSpecDBContext();
+
```
Comment (vicancy) [inline tools/RestfulApiWebService/OperationOnDB/Controllers/ApiFileController.cs]: Remove extra lines

File: tools/RestfulApiWebService/OperationOnDB/Controllers/ApiFileController.cs [hunk 28]
```diff
@@ -0,0 +1,75 @@
...
+        [HttpPost]
+        [Route("add")]
+        public string UploadFiles()
+        {
+            int uploadedFileNumber = 0;
+            string fileBasePath = HostingEnvironment.MapPath("~/uploads");
+
+            HttpFileCollection hfc = HttpContext.Current.Request.Files;
+
+            for(int count = 0; count < hfc.Count; count++)
```
Comment (vicancy) [inline tools/RestfulApiWebService/OperationOnDB/Controllers/ApiFileController.cs]: use foreach ?

File: tools/RestfulApiWebService/OperationOnDB/Controllers/ApiFileController.cs [hunk 29]
```diff
@@ -0,0 +1,75 @@
...
+            {
+                HttpPostedFile hpf = hfc[count];
+                if(hpf.ContentLength > 0)
+                {
+                    string FullPath = Path.Combine(fileBasePath, Path.GetFileName(hpf.FileName));
+                    if (!File.Exists(FullPath))
+                    {
+                        hpf.SaveAs(FullPath);
```
Comment (vicancy) [inline tools/RestfulApiWebService/OperationOnDB/Controllers/ApiFileController.cs]: no need to save, deserialize from `hpf.InputStream` directly. When changing to aspnet core, use `IFormFile` and read from `IFromFile.OpenReadStream` instead, referencing https://www.mikesdotnetting.com/article/288/uploading-files-with-asp-net-core-1-0-mvc

File: tools/RestfulApiWebService/OperationOnDB/Controllers/ApiFileController.cs [hunk 30]
```diff
@@ -0,0 +1,75 @@
...
+using Newtonsoft.Json;
+using OperationOnDB.Filters;
+using System.IO;
+using System.Web;
+using System.Web.Hosting;
+using System.Web.Http;
+
+namespace OperationOnDB.Controllers
+{
+    [ApiBasicAuthenticationAttribute("user", "user",
```
Comment (vicancy) [inline tools/RestfulApiWebService/OperationOnDB/Controllers/ApiFileController.cs]: so everyone knows the username and password after reading this line?

File: tools/RestfulApiWebService/OperationOnDB/Controllers/FileController.cs [hunk 31]
```diff
@@ -0,0 +1,70 @@
+﻿using DB.Dao;
```
Comment (vicancy) [inline tools/RestfulApiWebService/OperationOnDB/Controllers/FileController.cs]: duplicate with ApiFileController.cs?

File: tools/RestfulApiWebService/DB/Dao/XrefSpecDBContext.cs (hunk 19 above)
Comment (hechaochao) [inline tools/RestfulApiWebService/DB/Dao/XrefSpecDBContext.cs]: database access obejct

---
In reply to: [131036696](https://github.com/dotnet/docfx/pull/1906#discussion_r131036696) [](ancestors = 131036696)

File: tools/RestfulApiWebService/OperationOnDB/Filters/BasicAuthenticationAttribute.cs [hunk 32]
```diff
@@ -0,0 +1,37 @@
+﻿using System;
```
Comment (vicancy) [inline tools/RestfulApiWebService/OperationOnDB/Filters/BasicAuthenticationAttribute.cs]: duplicate one?

File: tools/RestfulApiWebService/OperationOnDB/Controllers/ApiFileController.cs [hunk 33]
```diff
@@ -0,0 +1,75 @@
...
+        private void SaveData(string path)
+        {
+
+            XRefMap xrefMap = YamlUtility.Deserialize<XRefMap>(path);
+            foreach (var spec in xrefMap.References)
+            {
+                XRefSpecObject xrefSpec = new XRefSpecObject();
+                xrefSpec.Uid = spec["uid"];
+                xrefSpec.HashedUid = MD5Encryption.CalculateMD5Hash(xrefSpec.Uid);
+                xrefSpec.XRefSpecJson = JsonConvert.SerializeObject(spec);
```
Comment (vicancy) [inline tools/RestfulApiWebService/OperationOnDB/Controllers/ApiFileController.cs]: use JsonUtility to save if client side uses JsonUtility to deserialize

**Comment by hechaochao** [comment]
> 

i will remove it

---
In reply to: [319839424](https://github.com/dotnet/docfx/pull/1906#issuecomment-319839424) [](ancestors = 319839424)

---
Refers to: tools/RestfulApiWebService/.gitignore:3 in dbe62d2. [](commit_id = dbe62d24491eca7c403a09cd4a5db26a532e09b8, deletion_comment = False)

File: tools/RestfulApiWebService/RestfulApiWebService/Controllers/QueryController.cs [hunk 34]
```diff
@@ -0,0 +1,81 @@
...
+                               .FirstOrDefaultAsync());
+            }
+
+            var uidList = await Task.WhenAll(getUidTasks);
+            foreach(var temp in uidList)
+            {
+                XRefSpec xf = null;
+                if (temp != null)
+                {
+                    xf = JsonConvert.DeserializeObject<XRefSpec>(temp.XRefSpecJson);
```
Comment (vicancy) [inline tools/RestfulApiWebService/RestfulApiWebService/Controllers/QueryController.cs]: no need to deserialize? use similar logic as GetByUid?

File: tools/RestfulApiWebService/DB/createDB.sql (hunk 24 above)
Comment (hechaochao) [inline tools/RestfulApiWebService/DB/createDB.sql]: it is no need i will remove it

---
In reply to: [131037163](https://github.com/dotnet/docfx/pull/1906#discussion_r131037163) [](ancestors = 131037163)

File: tools/RestfulApiWebService/OperationOnDB/Filters/BasicAuthenticationAttribute.cs (hunk 32 above)
Comment (hechaochao) [inline tools/RestfulApiWebService/OperationOnDB/Filters/BasicAuthenticationAttribute.cs]: one is mvc, another is for web api, there are a little differences

---
In reply to: [131039250](https://github.com/dotnet/docfx/pull/1906#discussion_r131039250) [](ancestors = 131039250)

File: tools/RestfulApiWebService/RestfulApiWebService/Controllers/QueryController.cs (hunk 34 above)
Comment (hechaochao) [inline tools/RestfulApiWebService/RestfulApiWebService/Controllers/QueryController.cs]: right, but the post method has go out of use

---
In reply to: [131039796](https://github.com/dotnet/docfx/pull/1906#discussion_r131039796) [](ancestors = 131039796)

File: tools/XRefService/DatabaseLib/Dao/XRefSpecDBContext.cs [hunk 35]
```diff
@@ -0,0 +1,15 @@
+﻿using DatabaseLib.Models;
```
Comment (vicancy) [inline tools/XRefService/DatabaseLib/Dao/XRefSpecDBContext.cs]: copyright headers

File: tools/XRefService/DatabaseLib/DatabaseLib.csproj [hunk 36]
```diff
@@ -0,0 +1,62 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (vicancy) [inline tools/XRefService/DatabaseLib/DatabaseLib.csproj]: Also use the new format csrpoj and reuse `..\..\src\Shared\base.props`, example: https://github.com/dotnet/docfx/blob/dev/tools/AzureMarkdownRewriterTool/AzureMarkdownRewriterTool.csproj

File: tools/XRefService/ManageXRefServiceDB/Controllers/FileController.cs [hunk 37]
```diff
@@ -0,0 +1,83 @@
...
+            _db.XRefSpecObjects.Add(new XRefSpecObject {HashedUid=MD5Encryption.CalculateMD5Hash(spec.Uid),
+                Uid=spec.Uid, XRefSpecJson =JsonUtility.Serialize(spec) });
+            _db.SaveChanges();
+            return Ok();
+        }
+
+        [HttpPost("upload")]
+        public async Task<IActionResult> UploadFile(ICollection<IFormFile> files)
+        {
+            //var uploads = Path.Combine(_environment.ContentRootPath, "uploads");
```
Comment (vicancy) [inline tools/XRefService/ManageXRefServiceDB/Controllers/FileController.cs]: remove useless comments

File: tools/XRefService/ManageXRefServiceDB/Controllers/FileController.cs [hunk 38]
```diff
@@ -0,0 +1,83 @@
...
+        }
+
+        private async Task SaveDataAsync(TextReader reader)
+        {
+            await Task.FromResult<int>(SaveData(reader));
+        }
+
+        private int SaveData(TextReader reader)
+        {
+
```
Comment (vicancy) [inline tools/XRefService/ManageXRefServiceDB/Controllers/FileController.cs]: remove extra empty line

File: tools/XRefService/ManageXRefServiceDB/Controllers/FileController.cs [hunk 39]
```diff
@@ -0,0 +1,83 @@
...
+                    }
+                }
+            }
+            await Task.WhenAll(tasks);
+            return StatusCode(200, "upload successfully");
+        }
+
+        private async Task SaveDataAsync(TextReader reader)
+        {
+            await Task.FromResult<int>(SaveData(reader));
```
Comment (vicancy) [inline tools/XRefService/ManageXRefServiceDB/Controllers/FileController.cs]: this is a fake async action...

File: tools/XRefService/ManageXRefServiceDB/ManageXRefServiceDB.csproj [hunk 40]
```diff
@@ -0,0 +1,50 @@
...
+  <PropertyGroup>
+    <TargetFramework>net461</TargetFramework>
+    <RuntimeIdentifier>win7-x86</RuntimeIdentifier>
+    <ApplicationIcon />
+    <OutputTypeEx>exe</OutputTypeEx>
+    <StartupObject />
+  </PropertyGroup>
+
+  <ItemGroup>
+    <Compile Remove="Filters\**" />
```
Comment (vicancy) [inline tools/XRefService/ManageXRefServiceDB/ManageXRefServiceDB.csproj]: duplicate items

File: tools/XRefService/ManageXRefServiceDB/ManageXRefServiceDB.csproj [hunk 41]
```diff
@@ -0,0 +1,50 @@
+﻿<Project Sdk="Microsoft.NET.Sdk.Web">
+
```
Comment (vicancy) [inline tools/XRefService/ManageXRefServiceDB/ManageXRefServiceDB.csproj]: reuse `..\..\src\Shared\base.props`

File: tools/XRefService/XRefService.sln [hunk 42]
```diff
@@ -0,0 +1,34 @@
+﻿
+Microsoft Visual Studio Solution File, Format Version 12.00
```
Comment (vicancy) [inline tools/XRefService/XRefService.sln]: remove, add projects to All.sln

File: tools/XRefService/ManageXRefServiceDB/Program.cs [hunk 43]
```diff
@@ -0,0 +1,21 @@
+﻿using System.IO;
+using Microsoft.AspNetCore.Hosting;
+
+namespace ManageXRefServiceDB
```
Comment (vicancy) [inline tools/XRefService/ManageXRefServiceDB/Program.cs]: naming: ManageXRefRestService ?

File: tools/XRefService/XRefService/Controllers/QueryController.cs [hunk 44]
```diff
@@ -0,0 +1,54 @@
...
+            string hashedUid = MD5Encryption.CalculateMD5Hash(uid);
+            var ut = await _db.XRefSpecObjects.Where(b => b.HashedUid == hashedUid)
+                               .Select(c => c.XRefSpecJson)
+                               .ToListAsync();
+            
+            return StatusCode(200, "[" + string.Join(",", ut) + "]");
+        }
+
+        [HttpGet]
+        [Route("extension/{uid}")]
```
Comment (vicancy) [inline tools/XRefService/XRefService/Controllers/QueryController.cs]: naming: don't think `extension` is a resource

File: tools/XRefService/DatabaseLib/Utility/MD5Encrytion.cs [hunk 45]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace DatabaseLib.Utility
+{
+    public class MD5Encryption
+    {
+        public static string CalculateMD5Hash(string input)
+        {
+            // step 1, calculate MD5 hash from input
+            MD5 md5 = MD5.Create();
+            byte[] inputBytes = Encoding.UTF8.GetBytes(input);
```
Comment (vwxyzh) [inline tools/XRefService/DatabaseLib/Utility/MD5Encrytion.cs]: `using System.Text.Encoding;`

File: tools/XRefService/DatabaseLib/Utility/MD5Encrytion.cs [hunk 46]
```diff
@@ -0,0 +1,24 @@
...
+            byte[] hash = md5.ComputeHash(inputBytes);
+
+            // step 2, convert byte array to hex string
+            StringBuilder sb = new StringBuilder();
```
Comment (vwxyzh) [inline tools/XRefService/DatabaseLib/Utility/MD5Encrytion.cs]: `new StringBuilder(32)`

File: tools/XRefService/XRefService/Controllers/QueryController.cs [hunk 47]
```diff
@@ -0,0 +1,54 @@
...
+        public async Task<IActionResult> GetByUidForExtension(string uid)
+        {
+            string hashedUid = MD5Encryption.CalculateMD5Hash(uid);
```
Comment (vwxyzh) [inline tools/XRefService/XRefService/Controllers/QueryController.cs]: remove

File: tools/XRefService/XRefService/Controllers/QueryController.cs [hunk 48]
```diff
@@ -0,0 +1,54 @@
...
+            var ut = await _db.XRefSpecObjects.Where(b => b.Uid.StartsWith(uid))
```
Comment (vwxyzh) [inline tools/XRefService/XRefService/Controllers/QueryController.cs]: Perf: full table scan in database.

File: tools/XRefService/DatabaseLib/Utility/MD5Encrytion.cs (hunk 45 above)
Comment (hechaochao) [inline tools/XRefService/DatabaseLib/Utility/MD5Encrytion.cs]: it's not a namespace it's a class

---
In reply to: [132074621](https://github.com/dotnet/docfx/pull/1906#discussion_r132074621) [](ancestors = 132074621)

File: tools/XRefService/ManageXRefServiceDB/Controllers/FileController.cs [hunk 49]
```diff
@@ -0,0 +1,83 @@
...
+            XRefMap xrefMap = YamlUtility.Deserialize<XRefMap>(reader);
+            foreach (var spec in xrefMap.References)
+            {
+                XRefSpecObject xrefSpec = new XRefSpecObject();
+                xrefSpec.Uid = spec["uid"];
+                xrefSpec.HashedUid = MD5Encryption.CalculateMD5Hash(xrefSpec.Uid);
+                xrefSpec.XRefSpecJson = JsonUtility.Serialize(spec);
+                _db.XRefSpecObjects.Add(xrefSpec);
+            }
+            _db.SaveChanges();
```
Comment (vwxyzh) [inline tools/XRefService/ManageXRefServiceDB/Controllers/FileController.cs]: use `SaveChangesAsync`

File: tools/XRefService/ManageXRefServiceDB/Controllers/FileController.cs [hunk 50]
```diff
@@ -0,0 +1,83 @@
...
+        {
+
+            XRefMap xrefMap = YamlUtility.Deserialize<XRefMap>(reader);
+            foreach (var spec in xrefMap.References)
+            {
+                XRefSpecObject xrefSpec = new XRefSpecObject();
+                xrefSpec.Uid = spec["uid"];
+                xrefSpec.HashedUid = MD5Encryption.CalculateMD5Hash(xrefSpec.Uid);
+                xrefSpec.XRefSpecJson = JsonUtility.Serialize(spec);
+                _db.XRefSpecObjects.Add(xrefSpec);
```
Comment (vwxyzh) [inline tools/XRefService/ManageXRefServiceDB/Controllers/FileController.cs]: what about updating?

File: tools/XRefService/ManageXRefServiceDB/Controllers/FileController.cs [hunk 51]
```diff
@@ -0,0 +1,83 @@
...
+
+        [HttpPost("add")]
+        public IActionResult Create([FromBody]XRefSpec spec)
+        {
+            if(spec == null)
+            {
+                return BadRequest();
+            }
+
+            _db.XRefSpecObjects.Add(new XRefSpecObject {HashedUid=MD5Encryption.CalculateMD5Hash(spec.Uid),
```
Comment (vwxyzh) [inline tools/XRefService/ManageXRefServiceDB/Controllers/FileController.cs]: format

File: tools/XRefService/XRefService.Common/Utility/MD5Encrytion.cs [hunk 52]
```diff
@@ -0,0 +1,27 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace XRefService.Common.Utility
```
Comment (vwxyzh) [inline tools/XRefService/XRefService.Common/Utility/MD5Encrytion.cs]: >Utility [](start = 29, length = 7)

naming, `Utilities`

File: tools/XRefService/XRefService/Controllers/QueryController.cs [hunk 53]
```diff
@@ -0,0 +1,64 @@
...
+        [HttpGet]
+        [Route("{uid}")]
+        public async Task<IActionResult> GetByUid(string uid)
+        {
+            string hashedUid = MD5Encryption.CalculateMD5Hash(uid);
+            var ut = await _db.XRefSpecObjects.Where(b => b.HashedUid == hashedUid)
+                               .Select(c => c.XRefSpecJson)
+                               .ToListAsync();
+            
+            return StatusCode(200, "[" + string.Join(",", ut) + "]");
```
Comment (vwxyzh) [inline tools/XRefService/XRefService/Controllers/QueryController.cs]: add content type.

**Comment by vicancy** [comment]
Track with https://github.com/dotnet/docfx/pull/2059 instead

**Pull request closed without merging**